### ISSUE TITLE
hint hash magic requires SV key

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#11962]
Migrated from [rt.perl.org#111000](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=111000) (status was 'resolved')

Searchable as RT111000$

**Comment by p5pRT** [comment]
### From zefram@fysh.org

#### Created by zefram@fysh.org



When magic on a hash causes magic to be applied to a new element in the
hash\, the details of the magic depend on whether the hash key was supplied
as an SV or as a raw C string\.  This is due to this part of hv\_common&#8203;:

&nbsp;                if \(keysv || is\_utf8\) \{
&nbsp;                    if \(\!keysv\) \{
&nbsp;                        keysv = newSVpvn\_utf8\(key\, klen\, TRUE\);
&nbsp;                    \}
&nbsp;                    if \(PL\_tainting\)
&nbsp;                        PL\_tainted = SvTAINTED\(keysv\);
&nbsp;                    keysv = sv\_2mortal\(newSVsv\(keysv\)\);
&nbsp;                    mg\_copy\(MUTABLE\_SV\(hv\)\, val\, \(char\*\)keysv\, HEf\_SVKEY\);
&nbsp;                \} else \{
&nbsp;                    mg\_copy\(MUTABLE\_SV\(hv\)\, val\, key\, klen\);
&nbsp;                \}

magic\_clearhint\(\) asssumes that the key is always recorded in SV form\,
and has an assertion about it&#8203;:

&nbsp;    assert\(mg\->mg\_len == HEf\_SVKEY\);

If the key is actually stored in plain string form then magic\_clearhint\(\)
crashes if the assertion wasn't active\.  The practical upshot of this
is that hints must be added via hv\_store\_ent\(\)\, not via hv\_store\(\)\.

magic\_sethint\(\)\, on the other hand\, has code to handle either form\.
I think magic\_clearhint\(\) should have the same logic\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl 5.14.2:

Configured by zefram at Mon Sep 26 19:44:21 BST 2011.

Summary of my perl5 (revision 5 version 14 subversion 2) configuration:
   
  Platform:
    osname=linux, osvers=2.6.26-2-686, archname=i386-linux-thread-multi
    uname='linux vigo.rous.org 2.6.26-2-686 #1 smp thu jan 27 00:28:05 utc 2011 i686 gnulinux '
    config_args='-des -Darchname=i386-linux -Dcccdlflags=-fPIC -Dccdlflags=-rdynamic -Dprefix=/home/zefram/usr/perl/perl_install/perl-5.14.2-i32-f52 -Dman1ext=1 -Dman3ext=3perl -Duselargefiles -Dusethreads -Uafs -Ud_csh -Uusesfio -Uusenm -Duseshrplib -Dusedevel -Uversiononly -Ui_db'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=undef, use64bitall=undef, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.4.5', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib/../lib /usr/lib/../lib /lib /usr/lib /usr/lib64
    libs=-lnsl -lgdbm -ldl -lm -lcrypt -lutil -lpthread -lc -lgdbm_compat
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lpthread -lc
    libc=/lib/libc-2.11.2.so, so=so, useshrplib=true, libperl=libperl.so
    gnulibc_version='2.11.2'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic -Wl,-rpath,/home/zefram/usr/perl/perl_install/perl-5.14.2-i32-f52/lib/5.14.2/i386-linux-thread-multi/CORE'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -L/usr/local/lib -fstack-protector'

Locally applied patches:
    


@INC for perl 5.14.2:
    /home/zefram/usr/perl/perl_install/perl-5.14.2-i32-f52/lib/site_perl/5.14.2/i386-linux-thread-multi
    /home/zefram/usr/perl/perl_install/perl-5.14.2-i32-f52/lib/site_perl/5.14.2
    /home/zefram/usr/perl/perl_install/perl-5.14.2-i32-f52/lib/5.14.2/i386-linux-thread-multi
    /home/zefram/usr/perl/perl_install/perl-5.14.2-i32-f52/lib/5.14.2
    .


Environment for perl 5.14.2:
    HOME=/home/zefram
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/zefram/usr/perl/perl_install/perl-5.14.2-i32-f52/bin:/home/zefram/usr/perl/util:/home/zefram/pub/i686-pc-linux-gnu/bin:/home/zefram/pub/common/bin:/usr/bin:/bin:/usr/local/bin:/usr/games
    PERL_BADLANG (unset)
    SHELL=/usr/bin/zsh

```
</details>

**Comment by p5pRT** [comment]
### From @cpansprout

There is a fix for this bug waiting on the sprout/misc\-post\-5\.16 branch\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

There is a fix for this bug waiting on the sprout/misc\-post\-5\.16 branch\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @cpansprout

Fixed in e33525913\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
